src/librustc_builtin_macros/asm.rs

```diff
@@ -147,8 +147,27 @@ fn parse_args<'a>(
                 }
             }
             ast::InlineAsmOperand::Sym { expr }
+        } else if allow_templates {
+            let template = p.parse_expr()?;
+            // If it can't possibly expand to a string, provide diagnostics here to include other
+            // things it could have been.
+            match template.kind {
+                ast::ExprKind::Lit(ast::Lit { kind: ast::LitKind::Str(..), .. }) => {}
+                ast::ExprKind::MacCall(..) => {}
```

Comment:
I don't think `MacCall` is ever reachable here since it would have been expanded already, but I'm not 100% sure.

Reply:
If that were true, we wouldn't need the call to `expr_to_spanned_string` in `expand_preparsed_asm` (which hasn't been called yet at this point). Removing `MacCall` from this match causes `concat!` or other macros to fail if passed in the second or subsequent template argument of an `asm!`. I have a test case for this.